File: server/app/database/schemas/order.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from enum import Enum


class OrderStatus(str, Enum):
    """
    Full customer-facing order lifecycle. Kept as a superset of every status
    string already written by existing code (pending/confirmed/processing/
    shipped/delivered/cancelled/refunded/paid/completed) plus the richer
    Amazon/Flipkart-style stages, so nothing already stored in MongoDB
    becomes invalid against this enum.
    """

    PENDING_PAYMENT = "pending_payment"
    PAYMENT_FAILED = "payment_failed"
    PENDING = "pending"  # legacy value, treated same as pending_payment
    PAID = "paid"  # legacy value written by order.py status-change endpoint
    CONFIRMED = "confirmed"
    PROCESSING = "processing"
    PACKED = "packed"
    READY_TO_SHIP = "ready_to_ship"
    SHIPMENT_CREATED = "shipment_created"
    PICKUP_SCHEDULED = "pickup_scheduled"
    SHIPPED = "shipped"
    IN_TRANSIT = "in_transit"
    OUT_FOR_DELIVERY = "out_for_delivery"
    DELIVERED = "delivered"
    COMPLETED = "completed"  # legacy value used by admin frontend; == delivered+settled
    CANCELLED = "cancelled"
    REFUNDED = "refunded"


class PaymentStatus(str, Enum):
    PENDING = "pending"
    AUTHORIZED = "authorized"
    CAPTURED = "captured"
    COMPLETED = "completed"  # legacy alias for captured, kept for compatibility
    FAILED = "failed"
    CANCELLED = "cancelled"
    REFUND_INITIATED = "refund_initiated"
    REFUNDED = "refunded"


# Statuses that represent a "final, no more automatic transitions" state.
TERMINAL_ORDER_STATUSES = {
    OrderStatus.DELIVERED,
    OrderStatus.COMPLETED,
    OrderStatus.CANCELLED,
    OrderStatus.REFUNDED,
}

# Statuses a customer is allowed to request cancellation from (i.e. before
# the order has shipped). Once shipped, cancellation must go through the
# return flow instead.
CANCELLABLE_STATUSES = {
    OrderStatus.PENDING_PAYMENT,
    OrderStatus.PENDING,
    OrderStatus.PAID,
    OrderStatus.CONFIRMED,
    OrderStatus.PROCESSING,
    OrderStatus.PACKED,
    OrderStatus.READY_TO_SHIP,
}


class OrderItemBase(BaseModel):
    product_id: str
    product_name: str
    product_image: str
    quantity: int
    size: str
    color: Optional[str] = None
    price: float
    total: float
    hsn_code: Optional[str] = None  # GST HSN/SAC code for invoicing


class OrderItemCreate(OrderItemBase):
    pass


class ShippingAddress(BaseModel):
    full_name: str
    phone: str
    address_line1: str
    address_line2: Optional[str] = None
    city: str
    state: str
    postal_code: str
    country: str = "India"


class OrderBase(BaseModel):
    user_id: str
    items: List[OrderItemBase]
    shipping_address: ShippingAddress
    subtotal: float
    discount: float = 0.0
    shipping_cost: float = 0.0
    tax: float = 0.0
    total_amount: float
    payment_method: str = "COD"
    notes: Optional[str] = None


class OrderCreate(OrderBase):
    pass


class OrderUpdate(BaseModel):
    status: Optional[OrderStatus] = None
    payment_status: Optional[PaymentStatus] = None
    tracking_number: Optional[str] = None
    notes: Optional[str] = None
    shipping_address: Optional[ShippingAddress] = None
    staff_assigned: Optional[str] = None
    warehouse_assigned: Optional[str] = None
    courier_name: Optional[str] = None


class Order(OrderBase):
    id: str = Field(alias="_id")
    order_number: str
    status: OrderStatus = OrderStatus.PENDING
    payment_status: PaymentStatus = PaymentStatus.PENDING
    tracking_number: Optional[str] = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    estimated_delivery: Optional[datetime] = None
    staff_assigned: Optional[str] = None
    warehouse_assigned: Optional[str] = None
    courier_name: Optional[str] = None

    class Config:
        populate_by_name = True

File: server/src/models/Order.js
const mongoose = require('mongoose');

const OrderItemSchema = new mongoose.Schema({
  product_id: { type: mongoose.Schema.Types.ObjectId, ref: 'Product', required: true },
  product_name: { type: String, required: true },
  product_image: { type: String },
  quantity: { type: Number, required: true, default: 1 },
  size: { type: String, required: true },
  color: { type: String },
  price: { type: Number, required: true },
  total: { type: Number, required: true },
  hsn_code: { type: String }
});

const ShippingAddressSchema = new mongoose.Schema({
  full_name: { type: String, required: true },
  phone: { type: String, required: true },
  address_line1: { type: String, required: true },
  address_line2: { type: String },
  city: { type: String, required: true },
  state: { type: String, required: true },
  postal_code: { type: String, required: true },
  country: { type: String, default: 'India' }
});

const OrderTimelineSchema = new mongoose.Schema({
  status: { type: String, required: true },
  changed_by: { type: String, required: true },
  role: { type: String, required: true },
  notes: { type: String },
  timestamp: { type: Date, default: Date.now }
});

const ShippingInfoSchema = new mongoose.Schema({
  shipment_id: { type: String },
  awb_code: { type: String },
  courier_name: { type: String },
  pickup_status: { type: String },
  tracking_url: { type: String },
  expected_delivery: { type: Date },
  shipment_status: { type: String }
});

const OrderSchema = new mongoose.Schema({
  order_number: { type: String, required: true, unique: true },
  user_id: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
  customer_email: { type: String, required: true },
  items: [OrderItemSchema],
  shipping_address: ShippingAddressSchema,
  subtotal: { type: Number, required: true },
  discount: { type: Number, default: 0.0 },
  shipping_cost: { type: Number, default: 0.0 },
  tax: { type: Number, default: 0.0 },
  total_amount: { type: Number, required: true },
  payment_method: { type: String, enum: ['COD', 'Razorpay'], default: 'COD' },
  payment_status: { 
    type: String, 
    enum: ['Pending', 'Authorized', 'Captured', 'Failed', 'Refunded'], 
    default: 'Pending' 
  },
  status: { 
    type: String, 
    enum: [
      'Pending Payment', 'Paid', 'Confirmed', 'Processing', 'Packed', 
      'Ready To Ship', 'Shipment Created', 'AWB Assigned', 'Pickup Requested', 
      'Picked Up', 'In Transit', 'Out For Delivery', 'Delivered', 
      'Completed', 'Cancelled', 'Refunded', 'Returned', 'RTO'
    ], 
    default: 'Pending Payment' 
  },
  coupon_code: { type: String },
  shipping: ShippingInfoSchema,
  timeline: [OrderTimelineSchema],
  notes: { type: String }
}, {
  timestamps: true
});

module.exports = mongoose.model('Order', OrderSchema);
